core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala

```diff
@@ -62,4 +88,31 @@ private[spark] trait ShuffleManager {
 
   /** Shut down this ShuffleManager. */
   def stop(): Unit
+
+  private[this] val shuffleToAttempts = new ConcurrentHashMap[Int, ConcurrentHashMap[Int, Int]]()
+
+  /**
+   * Register a stage attempt for the given shuffle, so we can clean up all attempts when
+   * the shuffle is unregistered
+   */
+  protected def addShuffleAttempt(shuffleId: Int, stageAttemptId: Int): Unit = {
+    shuffleToAttempts.putIfAbsent(shuffleId, new ConcurrentHashMap[Int, Int]())
+    shuffleToAttempts.get(shuffleId).putIfAbsent(stageAttemptId, stageAttemptId)
```

Comment:
Wait, why is the value of the map a map itself, if you're basically using it as a list? Maybe a `CopyOnWriteArraySet` or some other thread-safe collection would be more appropriate?

Or maybe just a plain `HashSet`, and use `shuffleToAttempts.get(shuffleId).synchronized` for thread-safety (since it should be rare for this to be called by the same shuffle concurrently, right?).